src/main/java/com/google/devtools/build/lib/rules/cpp/CppLinkAction.java

```diff
@@ -466,20 +484,117 @@ protected String getRawProgressMessage() {
   }
 
   /**
-   * Estimates resource consumption when this action is executed locally. During investigation we
-   * found linear dependency between used memory by action and number of inputs. For memory
-   * estimation we are using form C + K * inputs, where C and K selected in such way, that more than
-   * 95% of actions used less than C + K * inputs MB of memory during execution.
+   * Estimates resource consumption when this action is executed locally.
+   *
+   * <p>Because resource estimation for linking can be very costly (we need to inspect the size of
+   * the inputs, which means we have to expand a nested set and stat its files), we memoize those
+   * metrics. We do this to enable logging details about these computations after the linker has
+   * run, without having to re-do them.</p>
    */
-  static ResourceSet estimateResourceConsumptionLocal(OS os, int inputs) {
-    switch (os) {
-      case DARWIN:
-        return ResourceSet.createWithRamCpu(/* memoryMb= */ 15 + 0.05 * inputs, /* cpuUsage= */ 1);
-      case LINUX:
-        return ResourceSet.createWithRamCpu(
-            /* memoryMb= */ Math.max(50, -100 + 0.1 * inputs), /* cpuUsage= */ 1);
-      default:
-        return ResourceSet.createWithRamCpu(/* memoryMb= */ 1500 + inputs, /* cpuUsage= */ 1);
+  @VisibleForTesting
+  static class LocalResourcesEstimator implements LocalResourcesSupplier {
+    private final ActionExecutionContext actionExecutionContext;
+    private final OS os;
+    private final NestedSet<Artifact> inputs;
+
+    /** Container for all lazily-initialized details. */
+    private static class LazyData {
+      private int inputsCount;
+      private long inputsBytes;
+      private ResourceSet resourceSet;
+
+      LazyData(int inputsCount, long inputsBytes, ResourceSet resourceSet) {
+        this.inputsCount = inputsCount;
+        this.inputsBytes = inputsBytes;
+        this.resourceSet = resourceSet;
+      }
+    }
+
+    private LazyData lazyData = null;
+
+    public LocalResourcesEstimator(
+        ActionExecutionContext actionExecutionContext, OS os, NestedSet<Artifact> inputs) {
+      this.actionExecutionContext = actionExecutionContext;
+      this.os = os;
+      this.inputs = inputs;
+    }
+
+    /** Performs costly computations required to predict linker resources consumption. */
+    private LazyData doCostlyEstimation() {
+      int inputsCount = 0;
+      long inputsBytes = 0;
+      for (Artifact input : inputs.toList()) {
+        inputsCount += 1;
+        try {
+          FileArtifactValue value =
+              actionExecutionContext.getInputMetadataProvider().getInputMetadata(input);
+          if (value != null) {
+            inputsBytes += value.getSize();
+          } else {
+            logger.atWarning().log(
+                "Linker metrics: failed to get size of %s: no metadata", input.getExecPath());
+          }
+        } catch (Exception e) {
```

Comment:
let's not swallow Exceptions en bloc. Either only swallow `IOException` (worse) or re-throw it wrapped into `ExecException` (better)

Reply:
The reason for swallowing exceptions here was that the input sizes aren't used yet, so it seems too heavy-handed to fail hard if something goes wrong. The idea was to have this in place first to measure linker behavior by looking at the logs, and once we have data and a better model, we can put it in place (and then make this a hard failure once we have observed no failures).

I've changed the code to catch `IOException` (the broad `Exception` was a left-over from a previous draft) and added a TODO.

What do you think?

Reply:
I don't have a strong opinion here; I'd prefer failing hard because if `stat()` doesn't work on an input file, the action will very likely not succeed anyway (every input file must be checksummed so that the action checksum can be computed).

A mitigating factor is that this `IOException` only ever happens if I/O is needed to get the metadata, which AFAIU only ever happens during action input discovery, which is not something link actions do so arguably the way this exception is handled is only a theoretical concern.